pkg/antctl/raw/supportbundle/command.go

```diff
@@ -692,9 +702,151 @@ func processResults(resultMap map[string]error, dir string) error {
 	} else {
 		return err
 	}
 }
+
+func downloadControlleBundleFromKubernetes(ctx context.Context, antreaClientset antrea.Interface, k8sClient kubernetes.Interface, dir string) error {
+	var errors []error
+	tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return err
+	}
+	defer defaultFS.RemoveAll(tmpDir)
+
+	controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(ctx, "antrea-controller", metav1.GetOptions{})
+	if err == nil {
+		data, err := yaml.Marshal(controllerInfo)
+		if err == nil {
+			err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "controllerinfo"), data, 0644)
+			errors = append(errors, err)
+		}
+	}
+	errors = append(errors, err)
+	pods, err := k8sClient.CoreV1().Pods("kube-system").List(ctx, metav1.ListOptions{
+		ResourceVersion: "0",
+		LabelSelector:   "app=antrea,component=antrea-controller",
+	})
+	errors = append(errors, err)
+	if err == nil {
+		for _, pod := range pods.Items {
+			err = downloadPodLogs(ctx, k8sClient, "controller", pod.Namespace, pod.Name, k8s.GetPodContainerNames(&pod), dir, tmpDir)
+			errors = append(errors, err)
+		}
+	}
+	return utilerror.NewAggregate(errors)
+}
+
+func downloadAgentBundleFromKubernetes(ctx context.Context, antreaClientset antrea.Interface, k8sClient kubernetes.Interface, failedNodes []string, dir string) error {
+	agentInfoList, err := antreaClientset.CrdV1beta1().AntreaAgentInfos().List(ctx, metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		return err
+	}
+
+	agentInfoMap := map[string]v1beta1.AntreaAgentInfo{}
+	for _, agentInfo := range agentInfoList.Items {
+		agentInfoMap[agentInfo.Name] = agentInfo
+	}
+	pods, err := k8sClient.CoreV1().Pods("kube-system").List(ctx, metav1.ListOptions{
+		ResourceVersion: "0",
+		LabelSelector:   "app=antrea,component=antrea-agent",
+	})
+	if err != nil {
+		return err
+	}
+	failedNodeSet := sets.NewString(failedNodes...)
+	var errors []error
+	for _, pod := range pods.Items {
+		if !failedNodeSet.Has(pod.Spec.NodeName) {
+			continue
+		}
+		tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		defer defaultFS.RemoveAll(tmpDir)
+
+		if agentInfo, ok := agentInfoMap[pod.Spec.NodeName]; ok {
+			data, err := yaml.Marshal(agentInfo)
+			errors = append(errors, err)
+			if err == nil {
+				err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "agentinfo"), data, 0644)
+				errors = append(errors, err)
+			}
+		}
+		err = downloadPodLogs(ctx, k8sClient, "agent_"+pod.Spec.NodeName, pod.Namespace, pod.Name, k8s.GetPodContainerNames(&pod), dir, tmpDir)
+		errors = append(errors, err)
+
+	}
+	return utilerror.NewAggregate(errors)
+}
+
+func downloadPodLogs(ctx context.Context, k8sClient kubernetes.Interface, comp string, namespace string, podName string, containers []string, dir string, tmpDir string) error {
+	var errors []error
+	for _, containerName := range containers {
```

Comment:
it would be a bit more readable if you defined a closure to handle one container (similar idea as in a previous comment)

```go
downloadContainerLogs := func(containerName string) error {
        // ...
}
for _, containerName := range containers {
        if err := downloadContainerLogs(containerName); err != nil {
                errors = append(errors, err)
        }
}
```